File: python/src/multiagent_boilerplate/tools/builtin/save_plan.py
from __future__ import annotations

from ..types import ToolContext


class SavePlanTool:
    name = "save_plan"
    description = (
        "Persist the current plan to durable memory before spawning subagents, so a context "
        "truncation mid-run doesn't lose the strategy."
    )
    input_schema = {"type": "object", "required": ["plan"], "properties": {"plan": {}}}
    exposable = False

    async def execute(self, input_: dict, ctx: ToolContext) -> dict:
        await ctx.runtime.save_plan(input_["plan"])
        return {"saved": True}


save_plan_tool = SavePlanTool()

File: typescript/src/tools/builtin/savePlan.ts
import type { Tool } from "../types.js";

export const savePlanTool: Tool<{ plan: unknown }, { saved: true }> = {
  name: "save_plan",
  description:
    "Persist the current plan to durable memory before spawning subagents, so a context truncation mid-run doesn't lose the strategy.",
  inputSchema: {
    type: "object",
    required: ["plan"],
    properties: { plan: {} },
  },
  async execute(input, ctx) {
    await ctx.runtime.savePlan(input.plan);
    return { saved: true };
  },
};
